### PR TITLE
fix: Change `enum` behaviour in to emit values on compile stage

### PR DESCRIPTION
**Pull request by dj-stormtrooper**
To fix the compatibility with Pro agent. See more details about the problem [here](https://www.notion.so/64fff317ea754dc0a01a700dfb9c98c1).

### CODE REVIEW

**Review by Finesse — APPROVED** [review]
LGTM

Did you check this with Pro agent?

I'd wait for a maintainer approval before merging.

**Comment by dj-stormtrooper** [comment]
> Did you check this with Pro agent?

Not yet, I'll check

File: src/sources/notification_permissions.test.ts
Comment (dj-stormtrooper) [inline src/sources/notification_permissions.test.ts:8]: This test is broken in default branch as well, I fix it here for simplicity

Comment (dj-stormtrooper) [reply src/sources/notification_permissions.test.ts:8]: cc @bayotop because you've added other edge cases to this test, maybe you're interested in the latest Safari changes (16.5 is the latest on browserstack, but perhaps it changed earlier in 16 version, I don't have other minor versions to check)

File: src/index.ts
Comment (dj-stormtrooper) [inline src/index.ts:56]: @xnerhu this type will produce string union of enum values as you suggested [here ](https://github.com/fingerprintjs/BotD/pull/122#issuecomment-1528823679)

Values are exported as separate token `BotKindEnum`

Feel free to suggest other approaches

Comment (Finesse) [reply src/index.ts:56]: Is it necessary to change in this PR? Since it works, I'd not change unless necessary. But this is just my opinion. Both options are ok to me.

Comment (Finesse) [reply src/index.ts:56]: BTW this is a minor but breaking change. Before `BotKind` could be used as a runtime value, not it can't.

Comment (dj-stormtrooper) [reply src/index.ts:56]: Yes, it's definitely a breaking change, I tried to re-export this enum using internal `const enum` but it also violates `isolatedModule` rule 

I think this approach is the most convenient

Comment (dj-stormtrooper) [reply src/index.ts:56]: See the difference between these two enums (first one is with numeric values and the other with string values):
<details>
<summary>Screenshot</summary>
image
</details>

I suppose we had issues only with numeric enums, but there is a chance that something will go wrong when we change compiler settings (e.g. rollup plugins)

So I think it's safer to not to use runtime enums in this case, especially considering that they produc more verbose code after compilation

Comment (Finesse) [reply src/index.ts:56]: The string enum looks prone to the same problem (IIFE with no return value). It is not used by Pro agent directly or indirectly, so I think it's ok to leave it as it was.

> So I think it's safer to not to use runtime enums in this case

What do you suggest: An object, a const enum or something else?

Comment (dj-stormtrooper) [reply src/index.ts:56]: > What do you suggest: An object, a const enum or something else?

I mean to keep object enum + type as separate tokens

Comment (Finesse) [reply src/index.ts:56]: What about the breaking change?

Comment (dj-stormtrooper) [reply src/index.ts:56]: I think we could respect that in version change

Or I can try to rewrite internal implementation to `const enum` (to keep bundle small) but use separate enum  or public API (perhaps simple test will help to keep it in sync)

I can see implicit issue that we might meet in the future if we will use this enum indirectly:

```
if (BotKind.Value) {
  return 0
} else {
  return -1
}
```

In this case everything will be valid from the API contract perspective (no empty values) but the logic is broken. 

But if you consider this as low risk I 100% don't mind to revert last commit to keep this PR simpler and avoid breaking changes

Comment (Finesse) [reply src/index.ts:56]: > I think we could respect that in version change

What do you mean?

> Or I can try to rewrite internal implementation to `const enum` (to keep bundle small) but use separate enum or public API (perhaps simple test will help to keep it in sync)

Looks inconvenient to maintain. Do you suggest `const enum` to reduce the minified code size?

> I can see implicit issue that we might meet in the future if we will use this enum indirectly

I don't understand the issue. Could you please rephrase?

> But if you consider this as low risk I 100% don't mind to revert last commit to keep this PR simpler and avoid breaking changes

Let the maintainers decide. I understand the urge for consistency, but keeping a regular enum for `BotKind` seems to be the best option in this case.

Comment (dj-stormtrooper) [reply src/index.ts:56]: > What do you mean?

I mean to use major increment for example

> Looks inconvenient to maintain. Do you suggest const enum to reduce the minified code size?

Not only, I mean to keep this enum resistant to tree shaking (previous enums were used only indirectly, we didn't import them explicitly)

> I don't understand the issue. Could you please rephrase?

Sure, I mean these type of comparisons `BotKind.Value === 'value'` will silently become broken (because right now we only can track API contract in integration test but not the correct values

This is good a example of case when broken runtime enums are hard to notice:
https://github.com/fingerprintjs/BotD/blob/eac928d967c72032924de595a82850c0bf483005/src/detectors/window_size.ts#L5

`State.Success` will be equal to `undefined` and therefore the logic based on this condition will be affected, even if the API contract is valid (it will return `false` as expected in some cases)

> Let the maintainers decide. I understand the urge for consistency, but keeping a regular enum for BotKind seems to be the best option in this case

Ok, I agree

Comment (Finesse) [reply src/index.ts:56]: > I mean to use major increment for example

Ok. Bumping the major version just for this change is too much, I think, more breaking changes should be collected.

> This is good a example

Now I understand, thanks.

@xnerhu Your decision is needed here.

**Comment by Finesse** [comment]
@dj-stormtrooper Please confirm that this change solves the problem in Pro agent

**Comment by dj-stormtrooper** [comment]
> @dj-stormtrooper Please confirm that this change solves the problem in Pro agent

Yes, new integration tests are passed (I checked them without changes as well, they capture this error), and checked manually

But I'd check again on staging CDN before deploy just in case

**Review by Finesse — APPROVED** [review]
@dj-stormtrooper Could you please check Pro compatibility again just in case?

**Review by xnerhu — DISMISSED** [review]

**Comment by dj-stormtrooper** [comment]
> @dj-stormtrooper Could you please check Pro compatibility again just in case?

@Finesse it's all good

**Comment by Finesse** [comment]
@xnerhu Good to merge and publish

**Review by Finesse — APPROVED** [review]

**Pull request merged**
